### DISCUSSION TITLE
afl vs afl++ findings

### DISCUSSION THREAD

**Discussion by asarubbo** [AFLplusplus/AFLplusplus#1612] [General]
Hello,

this is the first time I'm using `afl++` while I used `afl` for years.

I recently managed to find two memory issues in [scanelf](https://bugs.gentoo.org/890028) with `afl` and I wanted to test the `afl++` behavior against scanelf.
So while `afl` find something after few minutes, `afl++` did not find anything after a day...I'm sure I have missed something but I can't understand what.

afl was launched in a way like:
```
ASAN_OPTIONS=""${ASAN_OPTIONS}",symbolize=0" afl-fuzz -i "${IN}" -o "${OUT}" -t "${TIMEOUT}" -f 0.elf -m none -M AFL0 /usr/bin/scanelf -s '*' @@
and the the other slave
```

afl++ was launched in a way like:
```
ASAN_OPTIONS=""${ASAN_OPTIONS}",symbolize=0" afl-fuzz -i "${IN}" -o "${OUT}" -t "${TIMEOUT}" -m none -M AFL0 /usr/bin/scanelf -s '*' @@
```
afl++ was tried with and without `-D`

The only big difference I can say is that with `afl` I compiled scanelf with the `afl-gcc` wrapper, while with `afl++` I used `afl-clang-fast`

Here is one of the afl++ slave that is running:
![Screenshot_20230108_201454](https://user-images.githubusercontent.com/4741819/211214558-3168a7dc-45c9-491e-a538-11f2e8259021.png)

**Comment by vanhauser-thc** [comment]
Can you try the crash you found with afl with the afl-clang-fast compiled binary? My guess would be that clang optimized out the issue.

**Comment by asarubbo** [comment]
Hi, I can try. However, tonight I did a run with the binary compiled with afl-gcc wrapper from afl++ (like I did with afl). It did not find anything and I can reproduce the crash with the afl++ instrumented binary and the old testcase.

**Comment by vanhauser-thc** [comment]
can you please retry with the most current afl++, like from this github state? what you are using is an older version.

**Comment by asarubbo** [comment]
Done, it looks like I get the same results:

![Screenshot_20230110_185620](https://user-images.githubusercontent.com/4741819/211639559-9cfd4cc7-7b9d-4b75-b81d-1eec69950790.png)

**Comment by vanhauser-thc** [comment]
how long does it take for afl to find the crash?
and can you please send me the whole directory of the afl instance finding the crash as a zip?
e.g. out/AFL99

and if you feel adventerous, you can try running afl++'s afl-fuzz with the `-Z` option and see if this makes it possible to find the crash

**Reply by asarubbo** [reply]
> how long does it take for afl to find the crash? and can you please send me the whole directory of the afl instance finding the crash as a zip? e.g. out/AFL99

Since you launched it on your own, do you still want those from me?
I launched afl++ with -Z (without -D). Waiting for results.

**Reply by vanhauser-thc** [reply]
not anymore, no. as you can see I found crashes with afl++. with a single seed, single instance, after 10 minutes.

(if that would not have been the case I would have used that out/AFLxx/ queue file as seed input for an afl++ runs that is the predecessor of one of the crashes (the crash name has id:xxxx and that id:xxxx is the id of the queue file on which the mutation was based that resulted in the crash)

**Comment by vanhauser-thc** [comment]
I let it run for a few minutes:
![image](https://user-images.githubusercontent.com/7396244/211765452-4b2ad964-e219-4ed4-9103-75b542d67287.png)
(note that the afl++ run was started a bit later, plus because of the different instrumentation it runs a bit slower, hence the afl instance has ~45% fuzz attempts at this point)

**Comment by asarubbo** [comment]
Seems that with [this](https://github.com/AFLplusplus/AFLplusplus/discussions/1613#discussioncomment-4655245) I can get crashes :)

**Reply by vanhauser-thc** [reply]
weird
